### PR TITLE
feat(git): cloneSubmodulesFilter

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds new option `cloneSubmodulesFilter` to allow selective cloning of submodules.

## Context

#33027

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Needs testing on at least one real repo

**Comment by rarkins** [comment]
I thought about converting `cloneSubmodules` from boolean to string list but decided it was too complex/risky. e.g. also need to change CLI, env, etc.

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -2856,6 +2856,14 @@ const options: RenovateOptions[] = [
     type: 'boolean',
     default: false,
   },
+  {
+    name: 'cloneSubmodulesFilter',
+    description:
+      'List of submodules names or patterns to clone when cloneSubmodules=true.',
+    type: 'array',
+    subType: 'string',
+    default: ['*'],
```
Comment (viceice) [inline lib/config/options/index.ts:2865]: `*`  only matches `something` but not `some/sub/module`
```suggestion
    default: ['**'],
```

File: lib/util/git/index.ts [hunk 2]
```diff
@@ -344,14 +345,24 @@ export async function getSubmodules(): Promise<string[]> {
   }
 }
 
-export async function cloneSubmodules(shouldClone: boolean): Promise<void> {
+export async function cloneSubmodules(
+  shouldClone: boolean,
+  cloneSubmodulesFilter: string[] = ['*'],
```
Comment (viceice) [inline lib/util/git/index.ts:350]:
```suggestion
  cloneSubmodulesFilter: string[] | undefined,
```

we shouldn't set a default value here, but require this arg to be passed, even when `undefined`

File: lib/util/git/index.ts [hunk 3]
```diff
@@ -344,14 +345,24 @@ export async function getSubmodules(): Promise<string[]> {
...
+): Promise<void> {
   if (!shouldClone || submodulesInitizialized) {
     return;
   }
   submodulesInitizialized = true;
   await syncGit();
   const submodules = await getSubmodules();
   for (const submodule of submodules) {
+    if (!matchRegexOrGlobList(submodule, cloneSubmodulesFilter)) {
```
Comment (viceice) [inline lib/util/git/index.ts:359]:
```suggestion
    if (cloneSubmodulesFilter && !matchRegexOrGlobList(submodule, cloneSubmodulesFilter)) {
```

**Review by viceice — APPROVED** [review]

**Pull request merged**